source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -477,12 +508,154 @@ Pass::Status AggressiveDCEPass::ProcessImpl() {
   // If any extensions in the module are not explicitly supported,
   // return unmodified.
   if (!AllExtensionsSupported()) return Status::SuccessWithoutChange;
-  // Process all entry point functions
+
+  // Eliminate Dead functions.
+  bool modified = EliminateDeadFunctions();
+
+  InitializeModuleScopeLiveInstructions();
+
+  // Process all entry point functions.
   ProcessFunction pfn = [this](ir::Function* fp) { return AggressiveDCE(fp); };
-  bool modified = ProcessEntryPointCallTree(pfn, get_module());
+  modified |= ProcessEntryPointCallTree(pfn, get_module());
+
+  // Process module-level instructions.
+  modified |= ProcessGlobalValues();
+
+  // Kill all dead instructions.
+  for (auto inst : to_kill_) {
+    context()->KillInst(inst);
+  }
+
+  // Cleanup all CFG including all unreachable blocks.
+  ProcessFunction cleanup = [this](ir::Function* f) { return CFGCleanup(f); };
+  modified |= ProcessEntryPointCallTree(cleanup, get_module());
+
   return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
 }
 
+bool AggressiveDCEPass::EliminateDeadFunctions() {
+  // Identify live functions first. Those that are not live
+  // are dead. ADCE is disabled for non-shaders so we do not check for exported
+  // functions here.
+  std::unordered_set<const ir::Function*> live_function_set;
+  ProcessFunction mark_live = [&live_function_set](ir::Function* fp) {
+    live_function_set.insert(fp);
+    return false;
+  };
+  ProcessEntryPointCallTree(mark_live, get_module());
+
+  bool modified = false;
+  for (auto funcIter = get_module()->begin();
+       funcIter != get_module()->end();) {
+    if (live_function_set.count(&*funcIter) == 0) {
+      modified = true;
+      EliminateFunction(&*funcIter);
+      funcIter = funcIter.Erase();
+    } else {
+      ++funcIter;
+    }
+  }
+
+  return modified;
+}
+
+void AggressiveDCEPass::EliminateFunction(ir::Function* func) {
+  // Remove all of the instruction in the function body
+  func->ForEachInst(
+      [this](ir::Instruction* inst) { context()->KillInst(inst); }, true);
+}
+
+bool AggressiveDCEPass::ProcessGlobalValues() {
+  // Remove debug and annotation statements referencing dead instructions.
+  // This must be done before killing the instructions, otherwise there are
+  // dead objects in the def/use database.
+  bool modified = false;
+  ir::Instruction* instruction = &*get_module()->debug2_begin();
+  while (instruction) {
+    if (instruction->opcode() != SpvOpName) {
+      instruction = instruction->NextNode();
+      continue;
+    }
+
+    if (IsTargetDead(instruction)) {
+      instruction = context()->KillInst(instruction);
+      modified = true;
+    } else {
+      instruction = instruction->NextNode();
+    }
+  }
+
+  std::vector<ir::Instruction*> annotations;
+  for (auto& inst : get_module()->annotations()) annotations.push_back(&inst);
+  std::sort(annotations.begin(), annotations.end(), DecorationSorter());
+  for (auto annotation : annotations) {
```

Comment:
Can you add a comment explaining what you are removing that will not be removed by KillInst?

Reply:
Ok